File: core/src/algo/kbo.rs
//! KBO: Knuth Bendix Ordering
//!
//! KBO is a reduction ordering widely used in automated theorem provers.
//! It is parametrized with a weight function and a precedence,
//! but here we provide default non-specified ones.

use {
    crate::{
        fnv,
        kernel_of_trust::{self as k, Expr, ExprView as EV, Var},
    },
    std::cmp::Ordering,
};

pub trait KBOParams {
    /// Precedence between symbols.
    fn prec(&self, a: &k::Symbol, b: &k::Symbol) -> Ordering;

    /// Weight of a symbol.
    fn weight(&self, _f: &k::Symbol) -> u16 {
        1
    }
}

/// Default KBO parameters.
#[derive(Debug, Copy, Clone)]
pub struct DefaultParams;

/// Compare the two expressions according to KBO.
pub fn compare(kbo: &mut dyn KBOParams, a: Expr, b: Expr) -> Option<Ordering> {
    use kbo_cmp::*;
    let mut st = State::new();
    let mut wb = Weight(0);
    tc_kbo(kbo, &mut st, &mut wb, &a, &b)
}

mod kbo_cmp {
    use super::*;

    use fnv::FnvHashMap as HM;

    pub struct State {
        /// Number of variables occurring only positively (in first term)
        pos_counter: u16,
        /// Number of variables occurring only negatively (in second term)
        neg_counter: u16,
        /// State for one variable, where we could occurrences in the LHS and RHS terms.
        var_balance: HM<k::Var, i32>,
    }

    #[derive(Clone, Copy, PartialEq, Eq, Ord, PartialOrd, Hash, Debug)]
    pub struct Weight(pub isize);

    impl State {
        pub fn new() -> Self {
            Self {
                pos_counter: 0,
                neg_counter: 0,
                var_balance: HM::default(),
            }
        }

        pub fn add_pos_var(&mut self, v: &k::Var) {
            let n = self.var_balance.get(&v).copied().unwrap_or(0);
            if n == 0 {
                self.pos_counter += 1;
            } else if n == -1 {
                self.neg_counter -= 1
            };
            self.var_balance.insert(v.clone(), n + 1);
        }

        pub fn add_neg_var(&mut self, v: &k::Var) {
            let n = self.var_balance.get(&v).copied().unwrap_or(0);
            if n == 0 {
                self.neg_counter += 1;
            } else if n == 1 {
                self.pos_counter -= 1
            };
            self.var_balance.insert(v.clone(), n - 1);
        }
    }

    /// Update term balance, weight balance, and check whether `t`
    /// contains the variable `s`.
    pub fn balance_weight(
        kbo: &mut dyn KBOParams,
        st: &mut State,
        wb: &mut Weight,
        t: &Expr,
        s: Option<&Var>,
        pos: bool,
    ) -> bool {
        match t.view() {
            EV::EBoundVar(..) => {
                if pos {
                    wb.0 += 1;
                } else {
                    wb.0 -= 1;
                }
                false
            }
            EV::EConst(c) => {
                let wc = kbo.weight(&c.name);
                assert!(wc > 0);
                if pos {
                    wb.0 += wc as isize;
                } else {
                    wb.0 -= wc as isize;
                }
                false
            }
            EV::EVar(v) => {
                if pos {
                    st.add_pos_var(v);
                } else {
                    st.add_neg_var(v);
                }
                if pos {
                    wb.0 += 1; // TODO: weight for variables?
                } else {
                    wb.0 -= 1;
                }
                if let Some(s) = s {
                    s == v
                } else {
                    false
                }
            }
            EV::EApp(f, a) => {
                let r1 = balance_weight(kbo, st, wb, f, s, pos);
                let r2 = balance_weight(kbo, st, wb, a, s, pos);
                r1 || r2
            }
            EV::ELambda(_ty, bod) => {
                if pos {
                    wb.0 += 1;
                } else {
                    wb.0 -= 1;
                }
                balance_weight(kbo, st, wb, bod, s, pos)
            }
            EV::EPi(..) | EV::EType | EV::EKind => false, // ignore types
        }
    }

    pub fn balance_weight_rec(
        kbo: &mut dyn KBOParams,
        st: &mut State,
        wb: &mut Weight,
        ts: &[&Expr],
        s: Option<&Var>,
        pos: bool,
    ) -> bool {
        let mut r = false;
        for t in ts {
            let rt = balance_weight(kbo, st, wb, t, s, pos);
            r = r || rt;
        }
        r
    }

    pub fn tc_kbo(
        kbo: &mut dyn KBOParams,
        st: &mut State,
        wb: &mut Weight,
        t1: &Expr,
        t2: &Expr,
    ) -> Option<Ordering> {
        if t1 == t2 {
            return Some(Ordering::Equal);
        }

        match (t1.view(), t2.view()) {
            (EV::EVar(x1), EV::EVar(x2)) => {
                st.add_pos_var(x1);
                st.add_neg_var(x2);
                None
            }
            (EV::EVar(x), _) => {
                st.add_pos_var(x);
                wb.0 += 1;
                let contains = balance_weight(kbo, st, wb, t2, Some(x), false);
                if contains {
                    Some(Ordering::Less)
                } else {
                    None
                }
            }
            (_, EV::EVar(y)) => {
                st.add_neg_var(y);
                wb.0 -= 1;
                let contains = balance_weight(kbo, st, wb, t1, Some(y), true);
                if contains {
                    Some(Ordering::Greater)
                } else {
                    None
                }
            }
            _ => {
                let (f1, ts1) = t1.unfold_app();
                let (f2, ts2) = t2.unfold_app();

                // precedences for heads
                let cmp_p = {
                    match (f1.view(), f2.view()) {
                        (EV::EConst(c1), EV::EConst(c2)) => Some(kbo.prec(&c1.name, &c2.name)),
                        _ => None,
                    }
                };

                balance_weight(kbo, st, wb, f1, None, true);
                balance_weight(kbo, st, wb, f2, None, false);

                let res_lex = if let Some(Ordering::Equal) = cmp_p {
                    tc_kbo_lex(kbo, st, wb, &ts1, &ts2)
                } else {
                    balance_weight_rec(kbo, st, wb, &ts1, None, true);
                    balance_weight_rec(kbo, st, wb, &ts2, None, false);
                    None
                };

                // in any case, check that no negative (resp. positive) var
                // occurs more than in the other side.
                let l_or_none = if st.neg_counter == 0 {
                    Some(Ordering::Less)
                } else {
                    None
                };

                let g_or_none = if st.neg_counter == 0 {
                    Some(Ordering::Greater)
                } else {
                    None
                };

                // now consider: weight, then precedence.
                if wb.0 > 0 {
                    g_or_none
                } else if wb.0 < 0 {
                    l_or_none
                } else if let Some(Ordering::Greater) = cmp_p {
                    g_or_none
                } else if let Some(Ordering::Less) = cmp_p {
                    l_or_none
                } else if let Some(Ordering::Equal) = cmp_p {
                    match res_lex {
                        Some(Ordering::Equal) => Some(Ordering::Equal),
                        Some(Ordering::Less) => l_or_none,
                        Some(Ordering::Greater) => g_or_none,
                        None => None,
                    }
                } else {
                    None
                }
            }
        }
    }

    /// Lexicographic order
    pub fn tc_kbo_lex(
        kbo: &mut dyn KBOParams,
        st: &mut State,
        wb: &mut Weight,
        mut t1: &[&Expr],
        mut t2: &[&Expr],
    ) -> Option<Ordering> {
        loop {
            if t1.len() == 0 && t2.len() == 0 {
                return Some(Ordering::Equal);
            } else if t1.len() == 0 {
                balance_weight_rec(kbo, st, wb, t2, None, false);
                return Some(Ordering::Less);
            } else if t2.len() == 0 {
                balance_weight_rec(kbo, st, wb, t1, None, true);
                return Some(Ordering::Greater);
            } else {
                match tc_kbo(kbo, st, wb, &t1[0], &t2[0]) {
                    Some(Ordering::Equal) => {
                        // continue further
                        t1 = &t1[1..];
                        t2 = &t2[1..];
                    }
                    res => {
                        // balance weights and return
                        balance_weight_rec(kbo, st, wb, t1, None, true);
                        balance_weight_rec(kbo, st, wb, t2, None, false);
                        return res;
                    }
                }
            }
        }
    }
}

mod impls {
    use super::*;

    impl KBOParams for DefaultParams {
        fn prec(&self, a: &k::Symbol, b: &k::Symbol) -> Ordering {
            // use the alphabetical order
            a.name().cmp(b.name())
        }
    }
}

File: src/core/KBO.ml
module K = Kernel
module E = K.Expr

type result =
  | Gt
  | Lt
  | Eq
  | Incomparable

type params = {
  var_as_cst: bool;
      (** if true, variables are just considered to be
      constants and are totally ordered in an unspecified but stable way.
      If false, then variables are incomparable and [a > b] implies that
      the multiset of variables of [a] contains the one of [b]. *)
  precedence: string -> string -> int;  (** Total order on constants *)
  weight: string -> int;
      (** Weight of constants. Must always be [>= 1]
      (we ignore the edge cases where one unary symbol can have weight 0
      for simplicity) *)
}

module KBO_ = struct
  type weight = int

  (* used to keep track of the balance of variables and the current
     weight difference *)
  type state = {
    mutable wb: weight;
    (* weight of LHS - weight of RHS *)
    mutable pos_counter: int;
    (* Number of variables occurring only positively (in first term) *)
    mutable neg_counter: int;
    (* Number of variables occurring only negatively (in second term) *)
    mutable balance: CCInt.t E.Tbl.t;
    (* State for one variable, where we could occurrences in the LHS and RHS terms. *)
    params: params;
  }

  let create params : state =
    {
      wb = 0;
      pos_counter = 0;
      neg_counter = 0;
      balance = E.Tbl.create 8;
      params;
    }

  (* add a positive variable *)
  let add_pos_var st var : unit =
    let n = E.Tbl.get_or st.balance var ~default:0 in
    if n = 0 then
      st.pos_counter <- st.pos_counter + 1
    else if n = -1 then
      st.neg_counter <- st.neg_counter - 1;
    E.Tbl.replace st.balance var (n + 1)

  (** add a negative variable *)
  let add_neg_var st var : unit =
    let n = E.Tbl.get_or st.balance var ~default:0 in
    if n = 0 then
      st.neg_counter <- st.neg_counter + 1
    else if n = 1 then
      st.pos_counter <- st.pos_counter - 1;
    E.Tbl.replace st.balance var (n - 1)

  let add_wb_ st x = st.wb <- st.wb + x

  let sub_wb_ st x = st.wb <- st.wb - x

  let incr_wb_ st = st.wb <- st.wb + 1

  let decr_wb_ st = st.wb <- st.wb - 1

  (* Update term balance, weight balance ([wb]), and check whether `t`
     contains the variable `s` *)
  let rec balance_weight st (e : E.t) (s : _ option) ~pos : bool =
    match E.view e with
    | E.E_kind | E.E_type | E.E_arrow _ | E.E_bound_var _ | E.E_box _ ->
      if pos then
        incr_wb_ st
      else
        decr_wb_ st;
      false
    | E.E_const (c, _) ->
      let wc = st.params.weight (K.Const.name c) in
      assert (wc > 0);
      if pos then
        add_wb_ st wc
      else
        sub_wb_ st wc;
      false
    | E.E_var v ->
      if pos then
        incr_wb_ st
      else
        decr_wb_ st;
      (* as if it were a constant of weight 1 *)
      if not st.params.var_as_cst then
        if pos then
          add_pos_var st e
        else
          add_neg_var st e;
      (match s with
      | Some s -> K.Var.equal s v
      | None -> false)
    | E.E_app (f, a) ->
      let r1 = balance_weight st f s ~pos in
      let r2 = balance_weight st a s ~pos in
      r1 || r2
    | E.E_lam (_, _ty, bod) ->
      if pos then
        incr_wb_ st
      else
        decr_wb_ st;
      balance_weight st bod s ~pos

  let balance_weight_rec (st : state) (es : E.t list) s ~pos : bool =
    List.fold_left
      (fun occurs e ->
        let r = balance_weight st e s ~pos in
        occurs || r)
      false es

  let rec tc_kbo st (t1 : E.t) (t2 : E.t) : result =
    if E.equal t1 t2 then
      Eq
    else (
      match E.view t1, E.view t2 with
      | E.E_var _, E.E_var _ ->
        add_pos_var st t1;
        add_neg_var st t2;
        Incomparable
      | E.E_var x, _ ->
        add_pos_var st t1;
        incr_wb_ st;
        let contains = balance_weight st t2 (Some x) ~pos:false in
        if contains then
          Lt
        else
          Incomparable
      | _, E.E_var x ->
        add_neg_var st t2;
        decr_wb_ st;
        let contains = balance_weight st t1 (Some x) ~pos:true in
        if contains then
          Gt
        else
          Incomparable
      | _ ->
        let f1, ts1 = E.unfold_app t1 in
        let f2, ts2 = E.unfold_app t2 in

        (* precedences for heads *)
        let cmp_p =
          match E.view f1, E.view f2 with
          | E.E_const (c1, _), E.E_const (c2, _) ->
            (* FIXME: if params.var_as_cst then also do variable cases *)
            let cmp =
              st.params.precedence (K.Const.name c1) (K.Const.name c2)
            in
            if cmp < 0 then
              Lt
            else if cmp = 0 then
              Eq
            else
              Gt
          | _ -> Incomparable
        in

        let (_ : bool) = balance_weight st f1 None ~pos:true in
        let (_ : bool) = balance_weight st f2 None ~pos:false in

        (* lexicographic comparison, if applicable *)
        let res_lex =
          match cmp_p with
          | Eq -> tc_kbo_lex st ts1 ts2
          | _ ->
            let (_ : bool) = balance_weight_rec st ts1 None ~pos:true in
            let (_ : bool) = balance_weight_rec st ts2 None ~pos:false in
            Incomparable
        in

        let l_or_none =
          if st.neg_counter = 0 then
            Lt
          else
            Incomparable
        in
        let g_or_none =
          if st.pos_counter = 0 then
            Gt
          else
            Incomparable
        in

        (* now consider: weight, then precedence *)
        if st.wb > 0 then
          g_or_none
        else if st.wb < 0 then
          l_or_none
        else if cmp_p = Gt then
          g_or_none
        else if cmp_p = Lt then
          l_or_none
        else if cmp_p = Eq then (
          match res_lex with
          | Eq -> Eq
          | Lt -> l_or_none
          | Gt -> g_or_none
          | Incomparable -> Incomparable
        ) else
          Incomparable
    )

  (* lexicographic order *)
  and tc_kbo_lex st (ts1 : E.t list) (ts2 : E.t list) : result =
    match ts1, ts2 with
    | [], [] -> Eq
    | [], _ ->
      let (_ : bool) = balance_weight_rec st ts2 None ~pos:false in
      Lt
    | _, [] ->
      let (_ : bool) = balance_weight_rec st ts1 None ~pos:true in
      Gt
    | t1 :: ts1, t2 :: ts2 ->
      (match tc_kbo st t1 t2 with
      | Eq -> tc_kbo_lex st ts1 ts2 (* continue on the rest of the lists *)
      | r ->
        (* balance weights and return r *)
        let (_ : bool) = balance_weight_rec st ts1 None ~pos:true in
        let (_ : bool) = balance_weight_rec st ts2 None ~pos:false in
        r)
end

let default_params () =
  { var_as_cst = false; weight = (fun _ -> 1); precedence = String.compare }

let compare ?(params = default_params ()) (e1 : E.t) (e2 : E.t) : result =
  let st = KBO_.create params in
  KBO_.tc_kbo st e1 e2

let lt ?params a b : bool = compare ?params a b = Lt

let gt ?params a b : bool = compare ?params a b = Gt
